thumbhash.cpp

```diff
@@ -0,0 +1,292 @@
+                                                           size_t nx,
+                                                           size_t ny,
+                                                           size_t w,
+                                                           size_t h)
+{
+    float dc = 0.0f;
+    FloatArray ac;
+    ac.reserve(nx * ny / 2);
+    float scale = 0.0f;
+    FloatArray fx(w, 0.0f);
+    for (size_t cy = 0; cy < ny; ++cy) {
+        size_t cx = 0;
+        while (cx * ny < nx * (ny - cy)) {
+            float f = 0.0f;
+            for (size_t x = 0; x < w; ++x) {
+                fx[x] = cos(PI / static_cast<float>(w) * static_cast<float>(cx) *
+                            (static_cast<float>(x) + 0.5f));
+            }
+            for (size_t y = 0; y < h; ++y) {
+                float fy = cos(PI / static_cast<float>(h) * static_cast<float>(cy) *
+                               (static_cast<float>(y) + 0.5f));
+                for (size_t x = 0; x < w; ++x) {
+                    f += channel[x + y * w] * fx[x] * fy;
+                }
+            }
+            f /= static_cast<float>(w * h);
+            if (cx > 0 || cy > 0) {
+                ac.push_back(f);
+                scale = std::max(std::abs(f), scale);
+            }
+            else {
+                dc = f;
+            }
+            cx += 1;
+        }
+    }
+    if (scale > 0.0) {
+        for (auto& ac_val : ac) {
+            ac_val = 0.5f + 0.5f / scale * ac_val;
+        }
+    }
+    return std::make_tuple(dc, ac, scale);
+}
+
+// This C++ thumbhash encode function is based on the rust reference
+// implementation found here:
+//
+// https://github.com/evanw/thumbhash/blob/main/rust/src/lib.rs
+//
+// We modified the logic in the following ways:
+//
+// - Make it work with OpenCV mat as input frame
+// - Handle images with or without an alpha channel
+// - Handle grayscale images
+// - Perform simple downscaling of large images. We don't need very many pixels
+//   to get a good hash.
+int thumbhash_encoder_encode(thumbhash_encoder e, const opencv_mat opaque_frame)
+{
+    auto frame = static_cast<const cv::Mat*>(opaque_frame);
+
+    size_t orig_w = frame->cols;
+    size_t orig_h = frame->rows;
+    size_t w = orig_w, h = orig_h;
+
+    // We don't need very many pixels to get a good hash. Downsample the image
+    // when its dimensions exceed the limit.
+    if (orig_w > MAX_DIMENSION || orig_h > MAX_DIMENSION) {
+        float aspect_ratio = static_cast<float>(orig_w) / orig_h;
+        if (orig_w > orig_h) {
+            w = MAX_DIMENSION;
+            h = static_cast<size_t>(w / aspect_ratio);
+        }
+        else {
+            h = MAX_DIMENSION;
+            w = static_cast<size_t>(h * aspect_ratio);
+        }
+    }
+
+    float row_ratio = static_cast<float>(orig_h) / h;
+    float col_ratio = static_cast<float>(orig_w) / w;
+
+    float avg_r = 0.0;
```

Comment:
it's a good idea. one hesitation i have is that i'd like to keep the structure of the code close to the rust impl in case we need to update it. and this change would move things around a bit more.

but since we added these other cases (3 channels and grayscale) to the code already, maybe we've already crossed that bridge.

here's what it looks like. let me know what you think:

https://github.com/discord/lilliput/commit/1f0d39c9c8f0a2262ac4ff5f2be70a772dcde615

Reply:
@walruscow ^

Reply:
I made the change as suggested.